src/server/raftkv.rs

```diff
@@ -335,32 +339,35 @@ impl<S: RaftStoreRouter> Engine for RaftKv<S> {
         let mut req = Request::default();
         req.set_cmd_type(CmdType::Snap);
 
-        ASYNC_REQUESTS_COUNTER_VEC.snapshot.all.inc();
-        let req_timer = ASYNC_REQUESTS_DURATIONS_VEC.snapshot.start_coarse_timer();
+        ASYNC_REQUESTS_COUNTER_VEC.may_flush(|m| m.snapshot.all.inc());
+        let req_timer = Instant::now_coarse();
 
         self.exec_read_requests(
             ctx,
             vec![req],
-            Box::new(move |(cb_ctx, res)| match res {
-                Ok(CmdRes::Resp(r)) => cb((
-                    cb_ctx,
-                    Err(invalid_resp_type(CmdType::Snap, r[0].get_cmd_type()).into()),
-                )),
-                Ok(CmdRes::Snap(s)) => {
-                    req_timer.observe_duration();
-                    ASYNC_REQUESTS_COUNTER_VEC.snapshot.success.inc();
-                    cb((cb_ctx, Ok(s)))
-                }
-                Err(e) => {
-                    let status_kind = get_status_kind_from_engine_error(&e);
-                    ASYNC_REQUESTS_COUNTER_VEC.snapshot.get(status_kind).inc();
-                    cb((cb_ctx, Err(e)))
+            Box::new(move |(cb_ctx, res)| {
+                ASYNC_REQUESTS_DURATIONS_VEC
+                    .may_flush(|m| m.snapshot.observe(duration_to_sec(req_timer.elapsed())));
```

Comment:
Maybe better to introduce an extention method for it to simplify the usage?

Reply:
Do  you mean implement something like `m.snapshot.observe_elapsed_sec(req_timer)`?

Reply:
Yes, or you can omit `_sec` because histogram by default use sec as the unit.